### PR TITLE
Introduction of Ember code test

### PR DESCRIPTION
**Pull request by aliciasedlock**
Tried to replicate as much from the Ruby side as possible, including comments and test names.

### CODE REVIEW

File: ember-test/.bowerrc [hunk 1]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (michaelpelletier) [inline ember-test/.bowerrc]: This is suuuuuuper pedantic but since the other folder is named ruby this should maybe just be ember instead of ember-test?

File: ember-test/.editorconfig [hunk 2]
```diff
@@ -0,0 +1,34 @@
+# EditorConfig helps developers define and maintain consistent
```
Comment (michaelpelletier) [inline ember-test/.editorconfig]: <3 that this is a thing.

File: ember-test/README.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
+# Tic Tac Toe: Ember Style
+
+Your job is to implement a simple game of tic-tac-toe. You have been provided with a test suite that must be passing before you submit your solution. These define, explicitly or implicitly, the required behavior of the system.
+
+As delivered, the game is missing these major parts that need implementing
+
+1. A game loop: A way for players to interact with the system and enter their moves
+2. A solution engine: A way for the game to let the players know who, if anyone, one the game.
```
Comment (michaelpelletier) [inline ember-test/README.md]: who, if anyone, "one" the game

File: ember-test/README.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+## Prerequisites
+
+You will need the following things properly installed on your computer.
+
+* [Git](http://git-scm.com/)
+* [Node.js](http://nodejs.org/) (with NPM)
+* [Bower](http://bower.io/)
+* [Ember CLI](http://www.ember-cli.com/)
+* [PhantomJS](http://phantomjs.org/)
```
Comment (michaelpelletier) [inline ember-test/README.md]: Does this section need to be specified if we are already having those things installed on the Candidates Laptop?

File: ember-test/bower.json [hunk 5]
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "ember-test",
+  "dependencies": {
+    "ember": "1.12.0",
```
Comment (michaelpelletier) [inline ember-test/bower.json]: Should this be running Ember 2.0?

File: ember-test/.bowerrc (hunk 1 above)
Comment (michaelpelletier) [inline ember-test/.bowerrc]: It occurs to me that this will also dick up sublime's "command t" when trying to open something in "test/"

File: ember-test/README.md (hunk 4 above)
Comment (aliciasedlock) [inline ember-test/README.md]: Hm, probably not. I'll keep it in until someone gives it a test run on the candidates laptop, and then we can remove it when we're sure it works. Sound good?

File: ember-test/bower.json (hunk 5 above)
Comment (aliciasedlock) [inline ember-test/bower.json]: Do we want 2.0 or 2.3?

File: ember-test/README.md (hunk 4 above)
Comment (michaelpelletier) [inline ember-test/README.md]: I forgot that people will have the option still of using their own laptops, so let's keep it.

File: ember-test/bower.json (hunk 5 above)
Comment (michaelpelletier) [inline ember-test/bower.json]: I thought 2.3 was still in beta. If it's final and there's a corresponding ember-cli version and everything, let's go to that.

Comment (aliciasedlock) [inline ember-test/bower.json]: Looks like 2.3 has graduated to release: http://emberjs.com/builds/#/release

File: ember-test/README.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+### Code Generators
+
+Make use of the many generators for code, try `ember help generate` for more details
+
+### Running Tests
+
+* `ember test`
+* `ember test --server`
+
+### Building
```
Comment (abmagil) [inline ember-test/README.md]: Should the "building" section come before the "running tests" section?  Will ember build itself if you run `ember test`?

File: ember-test/app/game/controller.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
+import Ember from 'ember';
+
+export default Ember.Controller.extend({
+  rows: [1,2,3], // defines number of rows our board has: done this way so we can send indicise to the action
```
Comment (abmagil) [inline ember-test/app/game/controller.js]: can this information be gleaned from the structure we're using as `board`?  Or is the comment indicating that we do it this way to save candidates the trouble of `row[x-1]` when accessing by index?

File: ember-test/bower.json (hunk 5 above)
Comment (abmagil) [inline ember-test/bower.json]: Talking with Michael, my thoughts are as follows (tl;dr at end)
I'd want to be able to present an Ember API that reflects what we do in production, without providing something that's EDGE or the latest release candidate.  Michael says that from 2.0 to 2.3, there weren't any real deprecations, just bug fixes.

**tl;dr** To me, that says 2.3 should be used

File: ember-test/tests/unit/game/controller-test.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+      assert.ok('player is alerted that move is invalid');
+    }
+  });
+});
+
+test('#entering a move updates the board', function(assert) {
+  assert.expect(1);
+
+  const controller = this.subject();
+  const expectedBoard = [
```
Comment (abmagil) [inline ember-test/tests/unit/game/controller-test.js]: Does this assertion preclude/suggest away from extracting a board model?

File: ember-test/bower.json (hunk 5 above)
Comment (aliciasedlock) [inline ember-test/bower.json]: :+1: we can always downgrade if we see the need.

File: ember-test/README.md (hunk 6 above)
Comment (aliciasedlock) [inline ember-test/README.md]: I can re-arrange this. This is what ember-cli provides by default so I didn't muck with it too much.

File: ember-test/app/game/controller.js (hunk 7 above)
Comment (aliciasedlock) [inline ember-test/app/game/controller.js]: I can add more comments for now, and if anyone thinks of a more clever way to set this up so we don't have to do this, I'm all for it.

File: ember-test/tests/unit/game/controller-test.js (hunk 8 above)
Comment (aliciasedlock) [inline ember-test/tests/unit/game/controller-test.js]: I didn't end up using a model because Ember models don't like arrays. They only support bools, strings, numbers, and dates, and I frankly did not feel like fighting/over-engineering a board. If there is a not crazy way to have a board model, I'm open to see what that looks like.

File: ember-test/README.md (hunk 6 above)
Comment (aliciasedlock) [inline ember-test/README.md]: Oh wait I'm awake now. Ember will build before running tests. The build command probably won't really get used, since there was watchers that rebuild the project upon changes and stuff.

Comment (abmagil) [inline ember-test/README.md]: :+1:

File: ember-test/tests/unit/game/controller-test.js (hunk 8 above)
Comment (abmagil) [inline ember-test/tests/unit/game/controller-test.js]: for realigans, Ember?!

File: ember-test/README.md (hunk 4 above)
Comment (aliciasedlock) [inline ember-test/README.md]: For some reason these comments won't get cleared as resolved but I updated the line to "You will need the following things properly installed on your computer. If you were given the developer interview laptop, these will already be installed."

File: ember-test/tests/unit/game/controller-test.js (hunk 8 above)
Comment (aliciasedlock) [inline ember-test/tests/unit/game/controller-test.js]: We would essentially need a `Board` model that has a `hasMany` relationship to a row model, which would need a `hasMany` relationship to a cell model, and then that could be a number, but I hate that it has to be so hyper-abstracted like that so I just..didn't.

If someone wanted to update the test to use a model system I wouldn't fight it. I would wonder if candidates would get really mixed up with that structure, and I'd rather them write code than figure out how models are related.

File: ember-test/bower.json (hunk 5 above)
Comment (sbatson5) [inline ember-test/bower.json]: I personally don't think it is going to matter much in this exercise.  Are we really going to care if the person doesn't use routable components or closure actions?

Comment (aliciasedlock) [inline ember-test/bower.json]: But if we're not on the latest version we don't look cool and hip.

File: ember-test/tests/unit/game/controller-test.js (hunk 8 above)
Comment (sbatson5) [inline ember-test/tests/unit/game/controller-test.js]: We have our own custom transform for `array` attr's in models, if we wanted to use that. However, my instinct initially would have been to not use models.  I don't see the benefit in this exercise.

Not to mention there are a lot of Ember developers who do _not_ use Ember-Data.  We shouldn't force it upon them.

Comment (aliciasedlock) [inline ember-test/tests/unit/game/controller-test.js]: That's a good point about Ember Data. I'm content with leaving the board stored as is. We can tweak after we run it a few times if we need to.

File: ember-test/.bowerrc (hunk 1 above)
Comment (michaelpelletier) [inline ember-test/.bowerrc]: Alicia said that it didn't like her calling it `ember` or `ember-cli` so let's break out our inner boston and call it `fahkin-embah-ked`

Only kinda joking.

Comment (abmagil) [inline ember-test/.bowerrc]: Bro...I know where you ember, bro.

File: ember-cli/package.json [hunk 9]
```diff
@@ -0,0 +1,37 @@
+{
+  "name": "ember-js",
+  "version": "0.0.0",
+  "description": "SoG test",
+  "private": true,
```
Comment (michaelpelletier) [inline ember-cli/package.json]: Does it matter that this isn't actually private?

File: ember-cli/package.json [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    "test": "tests"
+  },
+  "scripts": {
+    "start": "ember server",
+    "build": "ember build",
+    "test": "ember test"
+  },
+  "repository": "",
+  "engines": {
+    "node": ">= 0.10.0"
```
Comment (michaelpelletier) [inline ember-cli/package.json]: Do we need to be limited by this version of node?

File: ember-cli/package.json [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+  },
+  "author": "",
+  "license": "MIT",
+  "devDependencies": {
+    "broccoli-asset-rev": "^2.0.2",
+    "ember-cli": "0.2.5",
```
Comment (michaelpelletier) [inline ember-cli/package.json]: Old versions as I mentioned with my faceholes.

File: ember-cli/package.json (hunk 9 above)
Comment (aliciasedlock) [inline ember-cli/package.json]: Seemingly not, but I'll update it anyway.
